File: php/src/vaas/Authentication/ResourceOwnerPasswordGrantAuthenticator.php
<?php

namespace VaasSdk\Authentication;

use Amp\Http\Client\HttpClient;

class ResourceOwnerPasswordGrantAuthenticator extends TokenReceiver implements AuthenticatorInterface
{
    private readonly string $clientId;
    private readonly string $userName;
    private readonly string $password;

    /**
     * The authenticator for the resource owner password grant type if you have a client id, username and password.
     * This is the choice if you have registered yourself on https://vaas.gdata.de/login. In this case, the client id is `vaas-customer`.
     * @param string $clientId The client id
     * @param string $userName Your username or email
     * @param string $password Your password
     * @param string|null $tokenUrl The optional token url. Defaults to 'https://account.gdata.de/realms/vaas-production/protocol/openid-connect/token'
     * @param HttpClient|null $httpClient Your optional custom http client.
     */
    public function __construct(string $clientId, string $userName, string $password, ?string $tokenUrl = null, ?HttpClient $httpClient = null)
    {
        parent::__construct($tokenUrl, $httpClient);
        $this->clientId = $clientId;
        $this->userName = $userName;
        $this->password = $password;
    }

    protected function tokenRequestToForm(): string
    {
        return http_build_query([
            'client_id' => $this->clientId,
            'username' => $this->userName,
            'password' => $this->password,
            'grant_type' => 'password',
        ]);
    }
}

File: python/src/vaas/authentication/resource_owner_password_grant_authenticator.py
import time
from authlib.integrations.httpx_client import AsyncOAuth2Client

from .authenticator_interface import AuthenticatorInterface
from ..vaas_errors import VaasAuthenticationError


class ResourceOwnerPasswordGrantAuthenticator(AuthenticatorInterface):
    """Authenticator using Resource Owner Password Grant"""

    def __init__(self, client_id, user_name, password, token_endpoint, verify=True):
        self.client_id = client_id
        self.user_name = user_name
        self.password = password
        self.token_endpoint = token_endpoint
        self.verify = verify

        self._access_token = None
        self._expires_at = 0  # UNIX timestamp

    async def get_token(self):
        if self._access_token and time.time() < self._expires_at:
            return self._access_token

        async with AsyncOAuth2Client(self.client_id, verify=self.verify) as client:
            try:
                token_response = await client.fetch_token(
                    self.token_endpoint,
                    username=self.user_name,
                    password=self.password,
                    grant_type="password"
                )
                self._access_token = token_response["access_token"]
                expires_in = token_response.get("expires_in", 3600)
                self._expires_at = time.time() + expires_in
                return self._access_token
            except Exception as e:
                raise VaasAuthenticationError(e)
